### PR TITLE
Object-oriented and Functional paradigm descriptions

### PR DESCRIPTION
**Pull request by petermcneil**
- Written basic functional programming description.
- Written basic OOP programming description.
- Changed the thread description.
- Reformatted file with links etc.

### CODE REVIEW

File: Glossary.md [hunk 1]
```diff
@@ -1,19 +1,58 @@
...
+1. [Programming languages](#programming-languages)
+2. [Concurrency](#concurrency)
+3. [Client-server computing](#client-server-computing)
+
+Programming Languages
+---
+
+**Object-oriented Paradigm**
+
+Object-oriented programming solves computing concepts and problems by modelling objects. Objects have states and behaviours; for example, a person has a certain hair colour (state) and this person can change the colour(behaviour). Inheritance is part of the core functionality of objects in OO languages. Classes can have subclasses, which enables the subclass to inherit the methods and fields of its superclass. The core concept of inheritance is simple, but it can be used to model complex systems.  Objects are a demonstration a core tenant of OOP, encapsulation. This mechanism is used to enclose methods and state in bundles and to restrict access to them.
```
Comment (AlmasB) [inline Glossary.md]: Can we highlight keywords, so they stand out? For example, *Inheritance*, or **Inheritance**, whichever is appropriate.

Comment (AlmasB) [inline Glossary.md]: > Objects are a demonstration a core tenant of OOP, encapsulation. This mechanism is used to enclose methods and state in bundles and to restrict access to them. 

I think we can remove the first sentence for simplicity. So maybe just:

**Encapsulation** is used to enclose methods and state in bundles and to restrict access to them.

File: Glossary.md [hunk 2]
```diff
@@ -1,19 +1,58 @@
...
+3. [Client-server computing](#client-server-computing)
+
+Programming Languages
+---
+
+**Object-oriented Paradigm**
+
+Object-oriented programming solves computing concepts and problems by modelling objects. Objects have states and behaviours; for example, a person has a certain hair colour (state) and this person can change the colour(behaviour). Inheritance is part of the core functionality of objects in OO languages. Classes can have subclasses, which enables the subclass to inherit the methods and fields of its superclass. The core concept of inheritance is simple, but it can be used to model complex systems.  Objects are a demonstration a core tenant of OOP, encapsulation. This mechanism is used to enclose methods and state in bundles and to restrict access to them. 
+
+Using _Java_ as the example language
```
Comment (AlmasB) [inline Glossary.md:14]: Not sure this is a good example under OOP. I think your above example (person, hair colour) was much better. Can we continue with that example in code? 

Generally, giving an informal introduction first, followed by a practical example is a really good way to explain something.

File: Glossary.md [hunk 3]
```diff
@@ -1,19 +1,58 @@
...
+    }
+}
+```
 
-**Object-oriented Paradigm** - TODO
+-TODO: expand on this
 
-**Functional Paradigm** - TODO
+**Functional Paradigm**  
+Functional programming solves computing problems by chaining functions together. Functions are computations that are idempotent, similar to mathematical functions. They have referential transparency - that is a function always returns the same output for a given input. Typically there is no state and therefore no side-effects. Iterative loops are not allowed in functional programs, instead recursive calls are used.
```
Comment (AlmasB) [inline Glossary.md]: > idempotent

Will the majority of readers know what that word means? Can we rephrase this?

**Review by AlmasB — CHANGES_REQUESTED** [review]
Overall looks very good, please see attached comments.

File: Glossary.md (hunk 3 above)
Comment (petermcneil) [reply Glossary.md]: Fixed with a simpler explanation and mentioning idempotent.

File: Glossary.md (hunk 2 above)
Comment (petermcneil) [reply Glossary.md:14]: Example changed to a Person, with which the hair colour changes.

**Review by AlmasB — APPROVED** [review]
Nice PR, thanks. Also I've added you as collaborator, please check your notifications if you haven't done so yet.

**Pull request merged**